TeamCode/src/main/java/org/firstinspires/ftc/teamcode/OctoQuadLocalizer.java

```diff
@@ -0,0 +1,163 @@
+        octoquad.setLocalizerTcpOffsetMM_X((float) -(mmPerTick * PARAMS.parYTicks));
+        octoquad.setLocalizerTcpOffsetMM_Y((float) (mmPerTick * PARAMS.perpXTicks));
+
+        octoquad.setLocalizerImuHeadingScalar((float) PARAMS.angularScalar);
+
+        octoquad.setLocalizerVelocityIntervalMS(25);
+
+
+
+        // Reset the localization, IMU, and encoders
+        octoquad.resetEverything();
+
+        ElapsedTime timeout = new ElapsedTime();
+        OctoQuadFWv3.LocalizerStatus currentStatus = octoquad.getLocalizerStatus();
+        while (currentStatus != OctoQuadFWv3.LocalizerStatus.RUNNING) {
+            try {
+                Thread.sleep(10);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            if (timeout.seconds() > 3) {
+                Log.println(
+                        Log.WARN,
+                        "OctoQuadLocalizer",
+                        "init: Calibration timeout reached, OQ still not ready. OctoQuad reports " + currentStatus
+                );
+                RobotLog.addGlobalWarningMessage(
+                        "OctoQuad still not ready, timeout reached. OctoQuad reports " + currentStatus +
+                                " Continuing anyway, good luck!"
+                );
+                break;
+            }
+            currentStatus = octoquad.getLocalizerStatus();
+        }
+
+        octoquad.setLocalizerPose(
+                (int) DistanceUnit.MM.fromInches(currentPose.position.x),
+                (int) DistanceUnit.MM.fromInches(currentPose.position.y),
+                (float) currentPose.heading.toDouble()
+        );
+    }
+
+    @Override
+    public Pose2d getPose() {
+        return currentPose;
+    }
+
+    @Override
+    public void setPose(Pose2d pose) {
+        currentPose = pose;
+        octoquad.setLocalizerPose(
+                (int) DistanceUnit.MM.fromInches(currentPose.position.x),
+                (int) DistanceUnit.MM.fromInches(currentPose.position.y),
+                (float) currentPose.heading.toDouble()
+        );
+    }
+
+    @Override
+    public PoseVelocity2d update() {
+        OctoQuadFWv3.LocalizerDataBlock localizer = new OctoQuadFWv3.LocalizerDataBlock();
+        octoquad.readLocalizerData(localizer);
+        if (localizer.localizerStatus != OctoQuadFWv3.LocalizerStatus.RUNNING || !localizer.crcOk
+        ) {
+            Log.println(
+                    Log.WARN,
+                    "OctoQuadLocalizer",
+                    "update: Bad data recieved from OctoQuad." +
+                            " Localizer status: " + localizer.localizerStatus +
+                            " CRC OK: " + localizer.crcOk
+            );
+            // throw away bad data (don't change pose)
+            // have to return something for velocity; hopefully this won't cause a sudden jump? need to test
+            return new PoseVelocity2d(new Vector2d(0,0),0);
+        }
```

Comment:
I'd rather a tight retry loop here for 50ms instead of immediately returning bad data. Continuing the discussion from the initialization check loop, if there are common failures then it may be worth plumbing some signal into the action level to stop the robot.